test/box-luatest/gh_7974_force_recovery_bugs_test.lua

```diff
@@ -0,0 +1,189 @@
+local fio = require('fio')
+local server = require('luatest.server')
+local t = require('luatest')
+
+local g = t.group()
+
+-- Checks that force recovery works with snapshot containing no user spaces
+-- and an unknown request type.
+--
+-- Snapshot generation instruction:
+-- 1. Patch this line to encode an unknown request type (e.g., 777):
+-- luacheck: no max comment line length
+-- https://github.com/tarantool/tarantool/blob/cba4e20aa2600713b55bdaad6fd84396257835f2/src/box/xrow.c#L1218;
+-- 2. Build and run Tarantool, call `box.cfg`;
+-- 3. Bump LSN, e.g., `box.space._schema:insert{'foo', 'bar'}`;
+-- 4. Call `box.snapshot`.
+g.test_uknown_request_type_force_recovery = function()
+    local datadir = 'test/box-luatest/gh_7974_data/unknown_request_type'
+    local s = server:new(
+        {
+            alias = 'ok',
+            box_cfg = {force_recovery = true},
+            datadir = datadir,
+        })
+    s:start()
+    s:drop()
+
+    s = server:new(
+        {
+            alias = 'fail',
+            box_cfg = {force_recovery = false},
+            datadir = datadir,
+        })
+    s:start{wait_until_ready = false}
+    local log = fio.pathjoin(s.workdir, s.alias .. '.log')
+    t.helpers.retrying({}, function()
+        t.assert_not_equals(s:grep_log("can't initialize storage: " ..
```

Comment:
Nit: You could use the plain assert here:

```Lua
t.assert(s:grep_log(...))
```

Reply:
As per our Lua style guide[^1]:

> Avoid implicit casting to boolean in if conditions like `if x then` or `if not x then`. Such expressions will likely result in troubles with `[box.NULL]`(https://www.tarantool.io/ru/doc/latest/reference/reference_lua/box_null/). Instead of those conditions, use `if x ~= nil` then and `if x == nil` then.

[^1]: https://www.tarantool.io/en/doc/latest/dev_guide/lua_style_guide/#idioms-and-patterns

Reply:
`grep_log` never returns `box.NULL` (it returns nil or a string) so I don't see much point to be verbose here. We use plain `t.assert()` with `grep_log` in other places.

Reply:
It's a matter of consistency.